---
title: "Day 1 - Module 4a: Sequencing"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
    html_document:
        theme: readable
        toc: true
        toc_depth: 4
        toc_float: true
        number_sections: true
        fig_caption: true
        keep_md: true
---

<!---
library(rmarkdown)
render('Module2a_UNIXShell.Rmd')
--->

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>

# Workflow Overview

Let's begin with a high-level overview of a bulk RNA-seq workflow, starting with library preparation and ending in a list of differentially expressed genes. As we move through this workflow in the workshop, we will elaborate on each part, including highlighting good experimental design and best bioinformatics practices.

> INSERT WORKFLOW DIAGRAM HERE

# Library Preparation

The [Univeristy of Michigan Advanced Genomics Core (AGC)](https://brcf.medicine.umich.edu/cores/advanced-genomics/) performs a variety bulk RNA-seq library preps. Different library preparations enable different research questions.

- Poly-A:
- Total RNA:
- small RNA:
- 3' TagSeq (QuantSeq):

For details on recommended inputs see AGC's [webpage for input recommendations](https://brcf.medicine.umich.edu/cores/advanced-genomics/sample-requirements/).

# Sequencing

## Types

Typically AGC does paired-end 100bp sequencing because that is most economical on the NovaSeq 6000--their main sequencer.

In general, paired-end reads enable higher specificity alignments, as do longer read lengths.

## How Does It Work?

Illumina resources. Video?

## Results

Once sequencing is completed, AGC demultiplexes the reads in order to separate them by lab and sample. A cost-saving mechanism for sequencing is to pool many experiments in a sequencing run, but keep them distinguishable from one another by the barcodes.

# FASTQ Files

The files representing the raw sequenced reads are called FASTQ files. There is one for each sample, and one for each read pair (usually denoted by an `R1` and `R2` in the name). FASTQ files are simply text files with a particular structure. To see that structure, let's have a look at an example FASTQ that you downloaded as part of the workshop material.

```
$ cd workshop_data/fastqs

$ head -4 good_example_R1.fastq
```

In the above command, `head` is a UNIX command that lets us see the top lines of a file, in this case we've specified the top 4 lines (`-4`), but by default it's the top 10 lines.

The result is:

```
@A00437:266:HVFMYDSXX:1:1101:1181:1000 1:N:0:AGCCTATC+TGCGTAAC
TTTATTGTTGATGGTTATTTTTTGTTTATGGTTATTTTGTGTTTATGGTTATTTTTTGTTTATGGTTATTTTTTATTGATGGTTATTTTTTGTTTATGGTT
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFF
```

When we said earlier that FASTQ files have a particular structure, we mean that each read occurs in a quadruplet:

- Line 1 gives the read information, often including the sequencer and flowcell used, as well as the barcode for the sample.
- Line 2 gives the actual sequences that were determined by the sequencer (always AGCTs).
- Line 3 is a separator `+`.
- Line 4 is the Illumina per-base quality scoring coding. This is referred to as PHRED33.

## Base Quality Scores

It's important to note that the 4th line of a read is not really intended to be read by human beings. Instead, as we'll see shortly, there are programs that convert the encoding to a numerical score which is more interpretable by us.

A historical note, you may see programs refer to PHRED33 and PHRED64 scoring systems. The older PHRED64 scoring system is based on a different encoding and was used by older sequencers. It is safe to assume that any new data you receive from AGC is PHRED33. (DISCUSS: Say something about older SRA data? Is it usually indicated in the accession that a PHRED64 was used?)

## Paired-end FASTQs

When paired-end sequencing is performed, we briefly mentioned that read pairs are separated into `R1` files and `R2` files. These files will always have the same number of lines, and the quadruplets correspond to the same read. That is, the first 4 lines of each file refer to one read pair. As we will see later, aligners require that this pairing not be disturbed.
