---
title: "Day 1 - Module 6a: Reference Genomes"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
    html_document:
        theme: readable
        toc: true
        toc_depth: 4
        toc_float: true
        number_sections: true
        fig_caption: true
        keep_md: true
---

# Workflow Overview

Let's return to the workflow diagram we saw earlier.

INSERT WORKFLOW IMAGE

In the library preparation step we prepared our samples for sequencing. This involves creating many millions of fragments whose relationship to each other is lost in the process. These fragments are then sequenced, resulting in a digital representation of the fragments and confidence in the base calls. We then looked over the quality of the sequences and some characteristics about them (e.g. GC content, duplication, overrepresented sequences).

In order for analysis to proceed, we have to take the many millions of sequenced fragments and map them back to the genome / transcriptome. This will allow us to quantify the expression of genes, and investigate differential expression. First, we need something to map **onto**.

# What is in a reference genome?

A reference genome represents our collective understanding of the genomic sequence of an organism. It is important to note that they are not necessarily complete (e.g. determining the sequence around centromeres is difficult), and they are the ongoing product of the work of many researchers. That fact is illustrated by the notion of reference genome builds (e.g. hg19 and hg38).

For the purposes of RNA-seq analysis, the two most important parts of a reference genome are the **sequence** and the **gene model**.

## Sequence

At the very least, a reference genome consists of the genomic sequence. As an example, here are the beginning lines of the human genome build hg19:

```
>chrM
GATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCAT
TTGGTATTTTCGTCTGGGGGGTGTGCACGCGATAGCATTGCGAGACGCTG
GAGCCGGAGCACCCTATGTCGCAGTATCTGTCTTTGATTCCTGCCTCATT
CTATTATTTATCGCACCTACGTTCAATATTACAGGCGAACATACCTACTA
AAGTGTGTTAATTAATTAATGCTTGTAGGACATAATAATAACAATTGAAT
GTCTGCACAGCCGCTTTCCACACAGACATCATAACAAAAAATTTCCACCA
AACCCCCCCCTCCCCCCGCTTCTGGCCACAGCACTTAAACACATCTCTGC
CAAACCCCAAAAACAAAGAACCCTAACACCAGCCTAACCAGATTTCAAAT
TTTATCTTTAGGCGGTATGCACTTTTAACAGTCACCCCCCAACTAACACA
...
```

The genomic sequence is stored in a [FASTA](https://en.wikipedia.org/wiki/FASTA_format) file. These files are similar to FASTQ files in that they store sequence information, but their format is different, as you may have already noticed. FASTA files have only two types of lines: descriptions and sequences. Description lines begin with `>` and often denote chromosomes (as above). The description lines can have more formatting (as described [here](https://en.wikipedia.org/wiki/FASTA_format#NCBI_identifiers)). The sequence lines generally contain ACGTs as well as Ns to indicate unknown sequence (see [here](https://en.wikipedia.org/wiki/FASTA_format#Sequence_representation) for detailed code).

## Gene Model

For well characterized organisms, the gene model is also included in the form of a [GTF](https://uswest.ensembl.org/info/website/upload/gff.html) file. Here is an example of a GTF for the hg19 genome:

```
chr1	unknown	exon	11874	12227	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
chr1	unknown	exon	12613	12721	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
chr1	unknown	exon	13221	14409	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
chr1	unknown	exon	14362	14829	.	-	.	gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8568";
chr1	unknown	exon	14970	15038	.	-	.	gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8568";
chr1	unknown	exon	15796	15947	.	-	.	gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8568";
chr1	unknown	exon	16607	16765	.	-	.	gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8568";
chr1	unknown	exon	16858	17055	.	-	.	gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8568";
chr1	unknown	exon	17233	17368	.	-	.	gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8568";
chr1	unknown	exon	17369	17436	.	-	.	gene_id "MIR6859-3"; gene_name "MIR6859-3"; transcript_id "NR_107063_2"; tss_id "TSS25677";
```

At a glance we observe the following information:

- Chromosome location information including the chromosome, start, end, and strand.
- The type of feature, e.g. exon, CDS, start_codon, end_codon.
- The gene ID to which the feature belongs.
- The transcript ID to which the feature belongs.

In short, the GTF file encodes the gene and isoform structure for the genome, allowing us to map reads to genes and isoforms in order to perform differential expression analysis.

Some notes about GTF files:

- Depending on where a reference genome comes from, the gene and transcript IDs can look different. In the example above we see the gene IDs are gene symbols and the transcript IDs are RefSeq IDs from the RefSeq Consortium. GTF files from ENSEMBL or NCBI may have different identifiers (some examples are below).
- Each line of a GTF is a "feature". Features can be repeated as in the case where an exon belongs to many transcripts.

The GTF file format can result in files that look quite different from one another. Here is an example of a GTF file for the hg38 genome using ENSEMBL identifiers:

```
KN196472.1	HAVANA	exon	14223	14535	.	-	.	exon_id "ENSE00003761344.1"; exon_number "1"; gene_id "ENSG00000281764.1"; gene_name "RP11-399E6.2"; gene_status "KNOWN"; gene_type "processed_pseudogene"; havana_gene "OTTHUMG00000005711.2"; havana_transcript "OTTHUMT00000015628.2"; level "2"; ont "PGO:0000004"; tag "basic"; transcript_id "ENST00000630776.1"; transcript_name "RP11-399E6.2-001"; transcript_status "KNOWN"; transcript_type "processed_pseudogene"; tss_id "TSS131252";
KN196472.1	HAVANA	transcript	14223	14535	.	-	.	gene_id "ENSG00000281764.1"; gene_name "RP11-399E6.2"; gene_status "KNOWN"; gene_type "processed_pseudogene"; havana_gene "OTTHUMG00000005711.2"; havana_transcript "OTTHUMT00000015628.2"; level "2"; ont "PGO:0000004"; tag "basic"; transcript_id "ENST00000630776.1"; transcript_name "RP11-399E6.2-001"; transcript_status "KNOWN"; transcript_type "processed_pseudogene"; tss_id "TSS131252";
```

Note that the structure is similar, i.e. tab-delimited fields, but there is additional information included:

- The chromosomes use a different identifier.
- The second column indicates which genome annotation algorithm is responsible for the assignment (HAVANA is one such).
- The last field contains much more information about the feature than the RefSeq example we first saw.

# Where to get a reference genome?
