---
title: "Reference Genomes and Genome Indexing"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
    html_document:
        theme: readable
        toc: true
        toc_depth: 4
        toc_float: true
        number_sections: true
        fig_caption: true
        keep_md: true
---

> # Objectives
> *
> *
> *
> *
> *

# Differential Expression Workflow

As a reminder, our overall differential expression workflow is shown below. In this lesson we will set the stage for steps 5 and 6 by discussing reference genomes and genome indexing, which are integral to genome alignments and gene/isoform quantification.

| Step | Task |
| :--: | ---- |
| 1 | Experimental Design |
| 2 | Biological Samples / Library Preparation |
| 3 | Sequence Reads |
| 4 | Assess Quality of Raw Reads |
| **5** | **Splice-aware Mapping to Genome** |
| **6** | **Count reads associated with genes** |
| 7 | Test for DE Genes |

Up to this point, we have encountered sequenced reads in FASTQ format. The reads in these files have no connection to genome locations, and no connection to each other. In order to figure out how much a gene is expressed in a sample, let alone among groups of samples, we need to be able to quantify the gene expression by mapping reads to genes.

We need some idea of where the genes are, and what their sequence is.

# Reference Genome Basics

A reference genome represents our collective understanding of the genomic sequence of an organism. It is important to note that they are not necessarily complete, they do not capture the variability present in populations, and they are the ongoing product of the work of many researchers. That fact is illustrated by the notion of reference genome builds (e.g. hg19 and hg38) which are released periodically by the [Genome Reference Consortium](https://www.ncbi.nlm.nih.gov/grc).

For the purposes of RNA-seq analysis, the two most important parts of a reference genome are the **sequence** and the **gene model**.

## Sequence

At the very least, a reference genome consists of the genomic sequence. The following is the beginning of the reference sequence for the human mitochondrial chromosome for genome build hg19:

```
>chrM
GATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCAT
TTGGTATTTTCGTCTGGGGGGTGTGCACGCGATAGCATTGCGAGACGCTG
GAGCCGGAGCACCCTATGTCGCAGTATCTGTCTTTGATTCCTGCCTCATT
CTATTATTTATCGCACCTACGTTCAATATTACAGGCGAACATACCTACTA
AAGTGTGTTAATTAATTAATGCTTGTAGGACATAATAATAACAATTGAAT
GTCTGCACAGCCGCTTTCCACACAGACATCATAACAAAAAATTTCCACCA
AACCCCCCCCTCCCCCCGCTTCTGGCCACAGCACTTAAACACATCTCTGC
CAAACCCCAAAAACAAAGAACCCTAACACCAGCCTAACCAGATTTCAAAT
TTTATCTTTAGGCGGTATGCACTTTTAACAGTCACCCCCCAACTAACACA
...
```

Genomic sequence is stored in a [FASTA](https://en.wikipedia.org/wiki/FASTA_format) file. These files are similar to FASTQ files in that they store sequence information, but their format is a little different in a couple ways:

1. Records are separated by lines beginning with `>` instead of `@`.
2. Only the sequence is stored in a FASTA file, there is no notion of quality attached to the nucleotides.

## Gene Models

For well-characterized organisms, a gene model is often included as part of the reference genome in a [GTF (Gene Transfer Format)](https://uswest.ensembl.org/info/website/upload/gff.html) file. Here is an example of a GTF for the hg19 build of the human genome:

```
chr1	unknown	exon	11874	12227	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
chr1	unknown	exon	12613	12721	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
chr1	unknown	exon	13221	14409	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
chr1	unknown	exon	14362	14829	.	-	.	gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8568";
...
```

Each column of a GTF file has a particular meaning. Based on the above, some columns are clear, and others not. Here's the complete breakdown:

| Column | Description |
| :----: | ----------- |
| 1 | Chromosome |
| 2 | Source, e.g. ENSEMBL, HAVANA |
| 3 | Gene feature, e.g. exon, intron, mRNA, transcript |
| 4 | Start location, 1-based |
| 5 | End location, 1-based |
| 6 | Score |
| 7 | Strand |
| 8 | Frame, relating to codons |
| 9 | Attribute, a semicolon separated list of key/value pairs giving additional information about the feature. |

A GTF file representing a gene model gives information about the location of each part of a gene and identity information. We will see a little later that there are different identifiers depending on the source of the gene model.

# Getting Reference Genomes

The [Illumina iGenomes](https://support.illumina.com/sequencing/sequencing_software/igenome.html) resource is one of the easiest ways to download a reference genome. It has builds from ENSEMBL, NCBI, and UCSC (more on the similarities and differences later).

Reference genomes can be **very large**, depending on the organism, and so we will not download one to the Amazon instance we are using for this workshop. Indeed, it is not advised to download these references to your personal computer either. Instead, these should be downloaded on the server (your labs or Great Lakes) where you intend to do the RNA-seq analysis.

To do that, you would right click the genome build you want to download, and select "Copy link location".

![iGenomes image for copying link location](images/genome_copy_link.png)

Then on the remote server you would go to the directory you'd like to download the genome to and type:

```
$ wget http://igenomes.illumina.com.s3-website-us-east-1.amazonaws.com/Homo_sapiens/NCBI/GRCh38/Homo_sapiens_NCBI_GRCh38.tar.gz
```

After the download finishes (it may take a while as it is tens of GB large), you can unpack it with:

```
$ tar -xf Homo_sapiens_NCBI_GRCh38.tar.gz
```

# Details

We will now discuss some details that arise in the context of reference genomes and file formats.

* When describing the start and stop locations in the GTF, we mentioned that they are 1-based, what does that mean?
* On the iGenomes page there were different sources for the human genome (Ensembl, NCBI, and UCSC). What is the difference between the different sources? What is the difference between the different builds?
* When looking at the features in the GTF, it looked like there were two ways of identifying a gene (e.g. `gene_name "DDX11L1"; transcript_id "NR_046018";`). What are the different identifiers? How can we convert between them?

## Genomic Coordinates

There are two conventions for representing genomic coordinates. They are called 0-based and 1-based, and, unfortunately, different bioinformatic file types sometimes use different conventions. Let's break down the conventions with an illustration, and an example:

```
0-based, left-open

1 2 3 4 5 6 7 8 9
| | | | | | | | |
-A-C-G-T-A-C-G-C-
 | | | | | | | |
 1 2 3 4 5 6 7 8

1-based, closed
```

Where is the start and end of TAC above in 0-based coordinates? 1-based?

* 0-based: Start is 4, end is 7. Interval is (4, 7].
* 1-based: Start is 4, end is 6. Interval is [4, 6].

How do you calculate the length of the interval?

* 0-based: 7 - 4 = 3, in general: end - start.
* 1-based: 6 - 4 + 1 = 3, in general: end - start + 1.

Gene-level analyses may be somewhat robust to differences in 0-based and 1-based locations, but as soon as analyses focus on nucleotide-resolution features (e.g. CpG methylation or variant calling), off-by-one errors can be problematic.

## Gene Identifiers

```
chr1	unknown	exon	11874	12227	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
chr1	unknown	exon	12613	12721	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
chr1	unknown	exon	13221	14409	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
```

```
1	pseudogene	gene	11869	14412	.	+	.	gene_id "ENSG00000223972"; gene_name "DDX11L1"; gene_source "ensembl_havana"; gene_biotype "pseudogene";
1	processed_transcript	transcript	11869	14409	.	+	.	gene_id "ENSG00000223972"; transcript_id "ENST00000456328"; gene_name "DDX11L1"; gene_source "ensembl_havana"; gene_biotype "pseudogene"; transcript_name "DDX11L1-002"; transcript_source "havana";
1	processed_transcript	exon	11869	12227	.	+	.	gene_id "ENSG00000223972"; transcript_id "ENST00000456328"; exon_number "1"; gene_name "DDX11L1"; gene_source "ensembl_havana"; gene_biotype "pseudogene"; transcript_name "DDX11L1-002"; transcript_source "havana"; exon_id "ENSE00002234944";
```

| Source | Chromosome Style | Gene Name | Gene ID | Transcript ID |
| ------ | ---------------- | ------ | ------- | ------------- |
| ENSEMBL/NCBI | 1 | DDX11L1 | ENSG00000223972 | ENST00000456328 |
| UCSC | chr1 | DDX11L1 | DDX11L1 | NR_046018 |

# Creating a Genome Index for Alignments



---

These materials have been adapted and extended from materials created by the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
